cores/esp8266/core_esp8266_i2s.c

```diff
@@ -30,147 +30,278 @@
+// descriptor has the 'EOF' field set to 1.
 void ICACHE_FLASH_ATTR i2s_slc_isr(void) {
   uint32_t slc_intr_status = SLCIS;
   SLCIC = 0xFFFFFFFF;
   if (slc_intr_status & SLCIRXEOF) {
     ETS_SLC_INTR_DISABLE();
-    struct slc_queue_item *finished_item = (struct slc_queue_item*)SLCRXEDA;
+    tx_irqs++;
+    slc_queue_item_t *finished_item = (slc_queue_item_t *)SLCRXEDA;
     memset((void *)finished_item->buf_ptr, 0x00, SLC_BUF_LEN * 4);//zero the buffer so it is mute in case of underflow
-    if (i2s_slc_queue_len >= SLC_BUF_CNT-1) { //All buffers are empty. This means we have an underflow
-      i2s_slc_queue_next_item(); //free space for finished_item
+    if (tx->slc_queue_len >= SLC_BUF_CNT-1) { //All buffers are empty. This means we have an underflow
+      i2s_slc_queue_next_item(tx); //free space for finished_item
     }
-    i2s_slc_queue[i2s_slc_queue_len++] = finished_item->buf_ptr;
-    if (i2s_callback) i2s_callback();
+    tx->slc_queue[tx->slc_queue_len++] = finished_item->buf_ptr;
+    if (tx->callback) tx->callback();
+    ETS_SLC_INTR_ENABLE();
+  }
+  if (slc_intr_status & SLCITXEOF) {
+    ETS_SLC_INTR_DISABLE();
+    rx_irqs++;
+    slc_queue_item_t *finished_item = (slc_queue_item_t *)SLCTXEDA;
+    finished_item->owner = 1; // Or else RX just stops
+    finished_item->myid++;
+    i2s_slc_queue_append_item(rx, finished_item->buf_ptr);
+    if (rx->callback) rx->callback();
     ETS_SLC_INTR_ENABLE();
   }
 }
 
 void i2s_set_callback(void (*callback) (void)){
-    i2s_callback = callback;
+  tx->callback = callback;
+}
+
+void i2s_rx_set_callback(void (*callback) (void)){
+  rx->callback = callback;
 }
 
-void ICACHE_FLASH_ATTR i2s_slc_begin(){
-  i2s_slc_queue_len = 0;
+static void ICACHE_FLASH_ATTR _alloc_channel(i2s_state_t *ch) {
   int x, y;
-  
+
+  ch->slc_queue_len = 0;
   for (x=0; x<SLC_BUF_CNT; x++) {
-    i2s_slc_buf_pntr[x] = malloc(SLC_BUF_LEN*4);
-    for (y=0; y<SLC_BUF_LEN; y++) i2s_slc_buf_pntr[x][y] = 0;
-
-    i2s_slc_items[x].unused = 0;
-    i2s_slc_items[x].owner = 1;
-    i2s_slc_items[x].eof = 1;
-    i2s_slc_items[x].sub_sof = 0;
-    i2s_slc_items[x].datalen = SLC_BUF_LEN*4;
-    i2s_slc_items[x].blocksize = SLC_BUF_LEN*4;
-    i2s_slc_items[x].buf_ptr = (uint32_t)&i2s_slc_buf_pntr[x][0];
-    i2s_slc_items[x].next_link_ptr = (int)((x<(SLC_BUF_CNT-1))?(&i2s_slc_items[x+1]):(&i2s_slc_items[0]));
+    ch->slc_buf_pntr[x] = malloc(SLC_BUF_LEN*4);
+    for (y=0; y<SLC_BUF_LEN; y++) ch->slc_buf_pntr[x][y] = 0;
+
+    ch->slc_items[x].unused = 0;
+    ch->slc_items[x].owner = 1;
+    ch->slc_items[x].eof = 1;
+    ch->slc_items[x].sub_sof = 0;
+    ch->slc_items[x].datalen = SLC_BUF_LEN * 4;
+    ch->slc_items[x].blocksize = SLC_BUF_LEN * 4;
+    ch->slc_items[x].buf_ptr = (uint32_t*)&ch->slc_buf_pntr[x][0];
+    ch->slc_items[x].next_link_ptr = (x<(SLC_BUF_CNT-1))?(&ch->slc_items[x+1]):(&ch->slc_items[0]);
+    ch->slc_items[x].myid = 0;
+  }
+}
+#if 0
```

Comment:
Does this #if 0 block serve any purpose? Can it be removed? If it's useful for debug, can it be surrounded by a debug conditional instead?